docs/core/extensions/sslstream-best-practices.md

```diff
@@ -0,0 +1,127 @@
+
+### Custom Certificate Trust
+
+When encountering a certificate which was not issued by any of the certificate authorities trusted by the machine (including self-signed certificates), the default certificate validation procedure will fail. One possible way to resolve this is add the necessary issuer certificates to the machine trusted store. That, however, might affect other applications on the system and is not always possible.
+
+The alternative solution is to specify custom trusted root certificates via an <xref:System.Security.Cryptography.X509Certificates.X509ChainPolicy>.
+
+```csharp
+SslClientAuthenticationOptions clientOptions = new();
+
+clientOptions.CertificateChainPolicy = new X509ChainPolicy()
+{
+    TrustMode = X509ChainTrustMode.CustomRootTrust,
+    CustomTrustStore = {
+        customIssuerCert
+    }
+};
+```
+
+### Ignoring specific validation errors
+
+Consider an IoT device without a persistent clock. After powering on, the clock of the device would start many years in the past and, therefore, all certificates would be considered "not yet valid". The following code shows a validation callback implementation which ignores validity period violations.
+
+```csharp
+static bool CustomCertificateValidationCallback(
+    object sender,
+    X509Certificate? certificate,
+    X509Chain? chain,
+    SslPolicyErrors sslPolicyErrors)
+{
+    if (chain is null)
+    {
+        return false;
+    }
+
+    foreach (X509ChainStatus status in chain.ChainStatus)
+    {
+        if ((status.Status & ~X509ChainStatusFlags.NotTimeValid) != X509ChainStatusFlags.NoError)
+        {
+            return false;
+        }
+    }
+
+    return true;
+}
+```
+
+### Certificate pinning
+
+Another situation where custom certificate validation is necessary is when clients expect servers to use a specific certificate, or a certificate from a small set of known certificates. This practice is known as [certificate pinning](https://owasp.org/www-community/controls/Certificate_and_Public_Key_Pinning). The following code snippet shows a validation callback which checks that the server presents a certificate with a specific known public key.
+
+```csharp
+static bool CustomCertificateValidationCallback(
+    object sender,
+    X509Certificate? certificate,
+    X509Chain? chain,
+    SslPolicyErrors sslPolicyErrors)
+{
+    if (certificate == null)
```

Comment:
This sample implementation should probably still reject RemoteCertificateNameMismatch